### PR TITLE
Create search filters by interface

### PR DESCRIPTION
**Pull request by spinillos**
**What this PR does / why we need it**:

We need to extend the filters for the search in enterprise. It implies join tables that are not available in OSS, so to achieve this it creates a Filter interface that it's needed to implement any filter and add it dynamically.

The filter implements some functions like `JOIN`, `IN` or `WHERE`. These functions are added inside sql query to filter more information or join any table. You can take a look [here](https://github.com/grafana/grafana/pull/39843/files#diff-c6399cb5035dadc733601c3694ed48447d6abceab7cf2ffc288689f241e41f08R615).

After create the filter, you should add it [here](https://github.com/grafana/grafana/pull/39843/files#diff-844920b0c21adad3efd3b20e064386356ddb70e4b1f5208e209615918b6e14a0R47) and give it a name. The name is going to be the new filter's name.

The way to create a new filter is a simple `key=value` pair and are retrieved as an array of values. It means that we can do something like `filter=a` or `filter=a&filter=b&filter=c`.

**Which issue(s) this PR fixes**:

https://github.com/grafana/grafana-enterprise/issues/1870

**Special notes for your reviewer**:
We need to add filter as dynamic way to be able to join Enterprise tables.
The PR has breaking changes and it hardly depends of https://github.com/grafana/grafana/pull/39825 (UI changes)

### CODE REVIEW

**Review by Clarity-89 — APPROVED** [review]
Don't have much feedback on the code but I tested the behavior and it works as expected.

File: pkg/models/search_user_filter.go
Comment (AgnesToulet) [inline pkg/models/search_user_filter.go:63]:
```suggestion
	if !ok || len(params) == 0 {
		return nil
	}

	filter, err := f(params)
	if err != nil {
		log.Warnf("Cannot initialise the filter %s: %s", filterName, err)
		return nil
	}

	return filter
```

**Review by AgnesToulet — CHANGES_REQUESTED** [review]
Looks good but I noticed that this is breaking the current UI (the active users filter doesn't work anymore). I think the frontend change to fix this should be included in this PR.

**Comment by Clarity-89** [comment]
> Looks good but I noticed that this is breaking the current UI (the active users filter doesn't work anymore). I think the frontend change to fix this should be included in this PR.

I have a separate PR for that, which will update the front end.

File: pkg/models/search_user_filter.go
Comment (AgnesToulet) [inline pkg/models/search_user_filter.go:1]: I'm not sure this is the best location to put most of this code. I think a `filters` package in the `searchusers` service folder would be more appropriate (plus it would follow the same pattern as in the Enterprise code).

**Comment by AgnesToulet** [comment]
> 
> 
> > Looks good but I noticed that this is breaking the current UI (the active users filter doesn't work anymore). I think the frontend change to fix this should be included in this PR.
> 
> I'll have a separate PR for that, which will update the front end.

Maybe it's worth merging the two PRs then? I know we try to keep PRs small and it makes sense when we introduce changes in the backend that are simply unused until the frontend is updated but merging this when we know it's breaking the current behavior doesn't seem right.

**Comment by Clarity-89** [comment]
> > > Looks good but I noticed that this is breaking the current UI (the active users filter doesn't work anymore). I think the frontend change to fix this should be included in this PR.
> > 
> > 
> > I'll have a separate PR for that, which will update the front end.
> 
> Maybe it's worth merging the two PRs then? I know we try to keep PRs small and it makes sense when we introduce changes in the backend that are simply unused until the frontend is updated but merging this when we know it's breaking the current behavior doesn't seem right.

I think that's actually fine in this case. The milestone for the feature is 8.3 so there's little risk that this will be released without the front end counterpart, and it doesn't break the pipelines. The front end PR is a bit large and will require reviews from a different squad so that's why I'm a bit hesitant to merge it into this one.

File: pkg/models/search_user_filter.go
Comment (spinillos) [reply pkg/models/search_user_filter.go:1]: You are right. I started to work with is from the beginning inside models and I didn't noticed about that it's better and simple replicate the Enterprise code.

**Review by AgnesToulet — APPROVED** [review]
Code looks good! But I'd suggest to wait for the frontend PR to be approved before merging this one.

**Pull request merged**
